src/perturbopy/postproc/utils/spectra_generate_pulse.py

```diff
@@ -0,0 +1,373 @@
+"""
+Utils for ultrafast spectroscopy: pump pulse generation and post-processing
+"""
+
+import os
+import time
+import numpy as np
+import matplotlib.pyplot as plt
+import warnings
+from scipy import special
+
+from perturbopy.io_utils.io import open_hdf5, close_hdf5
+
+from matplotlib.animation import FuncAnimation
+
+from .memory import get_size
+from .timing import TimingGroup
+from .constants import energy_conversion_factor
+
+from . import spectra_plots
+
+
+def gaussian(x, mu, sig, hole_nband, elec_nband):
+    """
+    Gaussian function normalized to unity max occupation
+    """
+
+    return np.exp(-0.5 * ((x - mu) / sig)**2) / (hole_nband * elec_nband)
+
+
+def sigma_from_fwhm(fwhm):
+    """
+    Gaussian sigma: f(t) = A / ( sigma * sqrt(2pi)) exp(-(t-t0)^2 / (2sigma^2))
+    """
+
+    sigma = 1.0 / (2.0 * np.sqrt(2.0 * np.log(2.0))) * fwhm
+
+    return sigma
+
+
+def delta_occs_pulse_coef(t, dt, tw, sigma):
+    """
+    Additional occupation due to the pulse excitation.
+    Assuming the Gaussian pulse shape, the occupation is increasing in
+    time according to the error function.
+    """
+
+    numer = special.erf((t - tw / 2) / (np.sqrt(2) * sigma)) - \
+        special.erf((t - dt - tw / 2) / (np.sqrt(2) * sigma))
+
+    denom = special.erf(tw / 2 / (np.sqrt(2) * sigma))
+
+    return 0.5 * numer / denom
+
+
+def gaussian_excitation(pump_file, occs_amplitude, time_grid, time_window, pulse_fwhm, time_step,
+                        hole=False, finite_width=True):
+    """
+    The Gaussian pulse excitation for the Perturbo dynamics.
+    Pump pulses for electrons and holes must be dumped into separate HDF5 files.
+    In Perturbo, set `pump_pulse = .true.` and `pump_pulse_fname = ...` in pert.in.
+
+    In Perturbo, step_0 will be initialized with `init_boltz_dist` parameter,
+    the pump pulse will be applied starting from step_1.
+
+    Parameters
+    ----------
+
+    pump_pulse_file : h5py.File
+        The HDF5 file for the pump pulse excitation. Typically, `pump_pulse_Epump_...h5`.
+
+    occs_amplitude : np.ndarray
+        Electron or hole occupation amplitude: for each k-point, its _additional_ occupation in time
+        is modeled as Gaussian with the given FWHM and amplitude.
+
+    time_window : float
+        The time window for the pump pulse (fs). During this time window, the pump pulse will be active.
+
+    pulse_fwhm : float
+        The full width at half maximum of the pump pulse (fs).
+
+    time_step : float
+        Time step in fs for the pump pulse generation.
+        Note: the pump_time_step MUST match the one used in the dynamics run in Perturbo!
+
+    hole : bool, optional
+        If True, the pump pulse is for holes. Default is False (electrons).
+
+    finite_width : bool, optional
+        If True, the pulse is finite in time. If False, the pulse is a step function
+        and occs_amplitude will be set as initial occupation.
+    """
```

Comment:
Need to have line with the `Return` - what we expect to obtain from this function

Reply:
Good point! Done!